src/libs/Middleware/Reauthentication.js

```diff
@@ -0,0 +1,88 @@
+import lodashGet from 'lodash/get';
+import CONST from '../../CONST';
+import * as NetworkStore from '../Network/NetworkStore';
+import * as Network from '../Network';
+import * as Authentication from '../Authentication';
+import * as PersistedRequests from '../actions/PersistedRequests';
+import * as Request from '../Request';
+
+/**
+ * Reauthentication middleware
+ *
+ * @param {Promise} response
+ * @param {Object} request
+ * @param {Boolean} isFromSequentialQueue
+ * @returns {Promise}
+ */
+function Reauthentication(response, request, isFromSequentialQueue) {
+    return response
+        .then((data) => {
+            if (data.jsonCode === CONST.JSON_CODE.NOT_AUTHENTICATED) {
+                const credentials = NetworkStore.getCredentials();
+
+                // Credentials haven't been initialized. We will not be able to re-authenticate with the API.
+                const unableToReauthenticate = (!credentials || !credentials.autoGeneratedLogin || !credentials.autoGeneratedPassword);
+                if (unableToReauthenticate) {
+                    if (isFromSequentialQueue) {
+                        throw new Error('Missing credentials required for authentication');
+                    }
+
+                    Network.replayRequest(request);
+                    return;
+                }
+
+                // There are some API requests that should not be retried when there is an auth failure like
+                // creating and deleting logins. In those cases, they should handle the original response instead
+                // of the new response created by handleExpiredAuthToken.
+                const shouldRetry = lodashGet(request, 'data.shouldRetry');
+                if (!shouldRetry) {
+                    if (isFromSequentialQueue) {
+                        return data;
+                    }
+
+                    request.resolve(data);
+                    return;
+                }
+
+                // We are already authenticating
+                if (NetworkStore.getIsAuthenticating()) {
+                    if (isFromSequentialQueue) {
+                        // This can happen if we are already in the process of authenticating via the main queue and then
+                        // the sequential queue gets flushed (e.g. because we went offline while we were authenticating, queued some requests
+                        // and then came back online which triggers the sequential queue to flush).
+                        // I think the solution is to maybe only ever have one call to Authenticate happening at any time no matter who calls it.
```

Comment:
I think I agree with this solution but I'm not sure what that might imply.

Reply:
Yeah, I haven't quite worked it out, but I think it'd be something like...

- We only ever have one call to Authenticate happening at a time and we have a reference to that promise in the middleware
- If a request is made and that promise hasn't resolved yet we wait until it does by hooking into it and using it's response

I will caveat that I also haven't observed the case I'm describing happen, but did write a failing test that seems to imply it could happen like this:

- Request made while online returns 407 which triggers a call to `Authenticate` to get queued, but not run
- We go offline the same time the response is received but before we actually call `Authenticate` (not sure if this is possible)
- We make some requests while offline
- We get "stuck" `isAuthenticating` (again, not sure if it's possible)
- The Sequential Queue runs and the first request gets a 407
- We can't continue because we are "already authenticating"

This is why I had the instinct to throw an error - which would reveal if it's possible or not.

Reply:
Actually, the situation I am describing can be prevented by just checking to see if we are offline when handling a 407 and throwing an error if we are. That seems like a simpler solution so gonna try that.

Reply:
Ah, cool. In the end, if we still really want to log for this case, then I think we can just leave a comment that says it's there for a bit to ensure that nothing ever happens (because if it did happen, we would never know it, and it would lead to bugs that would be very hard to debug) and maybe we can go in later and remove the error.